src/beamlime/constructors/inspectors.py

```diff
@@ -37,11 +37,25 @@ def validate_annotation(annotation) -> Literal[True]:
 
 
 Product = TypeVar("Product")
-ProductType = Type[Product]
+_NewT = NewType("_NewT", int)
 
 
-def extract_underlying_product_type(product_type: ProductType) -> ProductType:
-    if hasattr(product_type, "__supertype__"):  # NewType
+class NewTypeMeta(type):
+    def __instancecheck__(cls, __instance: Any) -> bool:
+        return (
+            callable(__instance)
+            and hasattr(__instance, "__supertype__")
+            and _NewT.__module__ == __instance.__module__
+            and _NewT.__qualname__ == __instance.__qualname__
+        )
+
+
+class NewTypeProtocol(Generic[Product], metaclass=NewTypeMeta):
+    __supertype__: Type[Product]
```

Comment:
I don't understand this. Do product types now have to be built using `NewTypeProtocol`? I don't see this in the tests.

Why does `__instancecheck__` check whther `__instance` is callable?

Also, `__instance` shouldn't have any underscores.

Reply:
All `class`es are callable and `Union` also returns a `function` object so it's no use to check if it's `Callable` or a `function`.

+ I removed underscores...!

Reply:
If it's for static type checking, please consider making it a type guard: https://docs.python.org/3/library/typing.html#typing.TypeGuard

Reply:
That is so cool....! But it's only available from `py3.10`. I'll open an issue to use it. But still we need `NewTypeProtocol` at least I guess...? To do something like `is_newtype(val) -> TypeGuard[NewTypeProtocol]: ...`.